dhall-docs/src/Dhall/Docs/Core.hs

```diff
@@ -244,69 +262,122 @@ saveHtml inputAbsDir outputAbsDir packageName DhallFile {path = absFile, ..} = d
+    dirToDirsMap = Map.map removeHereDir $ go dirs initialMap
+      where
+        removeHereDir :: [Path Rel Dir] -> [Path Rel Dir]
+        removeHereDir = filter f
+          where
+            f :: Path Rel Dir -> Bool
+            f reldir = Path.parent reldir /= reldir
+
+        dirs :: [Path Rel Dir]
+        dirs = Map.keys dirToFilesMap
+
+        initialMap :: Map (Path Rel Dir) [Path Rel Dir]
+        initialMap = Map.fromList $ map (,[]) dirs
+
+        go :: [Path Rel Dir] -> Map (Path Rel Dir) [Path Rel Dir] -> Map (Path Rel Dir) [Path Rel Dir]
+        go [] m = m
+        go (d : ds) dirMap = go ds $ Map.adjust ([d] <>) (key $ Path.parent d) dirMap
+          where
+            key :: Path Rel Dir -> Path Rel Dir
+            key dir = if dir `Map.member` dirMap then dir else key $ Path.parent dir
+
+        -- toMap :: Path Rel Dir -> Map (Path Rel Dir) [Path Rel Dir]
+        -- toMap dir = Map.singleton dir $ Data.Set.filter f dirs
+        --   where
+        --     f :: Path Rel Dir -> Bool
+        --     f dir' = (dir == dir') ||
+        --         case FilePath.dropTrailingPathSeparator $ Path.fromRelDir currentDir of
+        --             "." -> False
+        --             _ -> f (Path.parent dir')
+
+    dirToDirsAndFilesMapAssocs :: [(Path Rel Dir, ([DhallFile], [Path Rel Dir]))]
+    dirToDirsAndFilesMapAssocs = Map.assocs $ Map.mapWithKey f dirToFilesMap
+      where
+        f :: Path Rel Dir -> [DhallFile] -> ([DhallFile], [Path Rel Dir])
+        f dir dhallFiles = case dirToDirsMap Map.!? dir of
+            Nothing -> fileAnIssue "Bug+with+indexes+creation"
+            Just dirs -> (dhallFiles, dirs)
+
+    toIndex :: (Path Rel Dir, ([DhallFile], [Path Rel Dir])) -> (Path Rel File, Text)
+    toIndex (indexDir, (dhallFiles, dirs)) =
+        (indexDir </> $(Path.mkRelFile "index.html"), Text.Lazy.toStrict $ Lucid.renderText html)
+      where
+        html = indexToHtml
+            indexDir
+            (map (\DhallFile{..} -> (stripPrefix $ addHtmlExt path, mType)) dhallFiles)
+            (map stripPrefix dirs)
+            DocParams { relativeResourcesPath = resolveRelativePath indexDir, packageName }
+
+        stripPrefix :: Path Rel a -> Path Rel a
+        stripPrefix relpath =
+            if Path.toFilePath relpath == Path.toFilePath indexDir then relpath
+            else Data.Maybe.fromMaybe (fileAnIssue "Bug+with+stripPrefix")
+                $ Path.stripProperPrefix indexDir relpath
+
+addHtmlExt :: Path Rel File -> Path Rel File
+addHtmlExt relFile =
+    Data.Maybe.fromMaybe (fileAnIssue "Bug+with+addHtmlExt") $ Path.addExtension ".html" relFile
+
+-- | If you're wondering the GitHub query params for issue creation:
+-- https://docs.github.com/en/github/managing-your-work-on-github/about-automation-for-issues-and-pull-requests-with-query-parameters
+fileAnIssue :: Text -> a
+fileAnIssue titleName =
+    error $ "\ESC[1;31mInternal Error\ESC[0mOops!\n\nThis should not happen.\n" <>
```

Comment:
```suggestion
    error $ "\ESC[1;31mError\ESC[0mDocumentation generator bug\n\nThis should not happen.\n" <>
```

I also suggest using an error text similar to the `Dhall.Util.internalError` utility:

https://hackage.haskell.org/package/dhall-1.33.1/docs/src/Dhall.Syntax.html#internalError

In particular, you probably want to avoid suggesting a title for the user's issue; the user is more likely to generate a useful title on their own.